Added middle C octave changer & Refactored some parts

ddxtanx opened the pull request:
PR to add middle C octave setter (here: https://github.com/walmik/scribbletune/issues/28) Also, while going through some code, it was formatted fairly weirdly so I refactored certain portions.

ddxtanx commented:
Added second commit due to oversight on the transposition function.

walmik commented:
Thanks for addressing an open issue!

I just got around taking a look at this PR, I ll look into it and get back with comments.

walmik commented on README.md:
````diff
@@ -135,6 +135,15 @@ Import the 3 MIDI files thus generated into your favorite music creation softwar
+### Changing Middle C
+Sometimes Middle C is not C4, but a C in a different octave. To set the octave for middle C use
+```js
+scribble.setMiddle(octave)
````
Rename the method to `setMiddleC` and give a full example of how to use this feature where a clip is created with a new middle C

walmik commented on src/midi.js:
```diff
@@ -3,7 +3,14 @@
+ * @param {Integer} octaveIndex		The new octave for middle C.
+ */
+function setMiddle(octaveIndex){
+	transposition = octaveIndex-4;
```
Add a space on the either side of the hypen. `octaveIndex - 4;` instead of `octaveIndex-4;`

walmik commented on src/midi.js:
```diff
@@ -22,19 +29,26 @@ const midi = (notes, fileName) => {
+			//Get the octave identifier (2 for a2, 5 for e5)
+			var oct = parseInt(note.slice(1,note.length));
+			//Parse the octave into an integer
+			oct+=transposition;
```
Add a space on the either side of `+=`.
`oct += transposition;` instead of `oct+=transposition;`

walmik commented on src/midi.js:
```diff
@@ -22,19 +29,26 @@ const midi = (notes, fileName) => {
+			oct+=transposition;
+			//Transpose the octave
+			noteObj.note = note[0]+oct.toString();
+			console.log(note);
```
Remove console.log

walmik commented on src/index.js:
```diff
@@ -1 +1,16 @@
+function setMiddle(octaveIndex){
+    midi.setMiddle(octaveIndex);
+    //Trans is the positive octave transposition
+    //If trans is 1, all notes will be shifted up one octave
```
Add a space after `//` for the above 2 lines, so that it look like
```
// Trans is the positive octave transposition
// If trans is 1, all notes will be shifted up one octave
```

walmik commented on src/midi.js:
```diff
@@ -3,7 +3,14 @@
 const jsmidgen = require('jsmidgen');
-
+var transposition = 0;
+/**
```
Add an empty line before declaring this function.

walmik commented on src/index.js:
```diff
@@ -1 +1,16 @@
+const pattern = require('./pattern');
+const midi = require('./midi');
+const scales = require('./modes');
+function setMiddle(octaveIndex){
```
At a interface level I m ok to do this `scribble.setMiddleC(Number)` but lets try to come up with a better way to put it together at the code level. Defining `setMiddle` in 2 different files seems counterintuitive. 

I ll give this some more thought too.

ddxtanx replied:
The only reason I had to declare it in 2 files was to expose the function to the user. I had to provide functionality to change the notes in midi.js, while also having it available to the user as scribble.setMiddle().

ddxtanx replied:
Sorry if it's a bit redundant, it's the only way I could think to do it.

walmik replied:
No worries, your underlying idea is certainly good, we just need to come up with some improvements to it.

walmik commented on src/midi.js:
```diff
@@ -3,7 +3,14 @@
+ * Takes an integer and transposes all notes to a different middle C octave.
+ * @param {Integer} octaveIndex		The new octave for middle C.
+ */
+function setMiddle(octaveIndex){
```
Throw an error in case user send in invalid value for `octaveIndex `.

ddxtanx replied:
What values would be invalid?

walmik replied:
Any non Numerical value. The objective is to cover edge cases. Knowingly or unknowingly, someone may call this method something like `scribble.setMiddleC('c3')` 

There are folks who script stuff that decides root notes, octaves, patterns etc randomly. For some reason that script ends up calling scribble.setMiddleC with something completely unexpected value, like `scribble.setMiddleC(true)` OR  `scribble.setMiddleC('D#')` then our code must let them know that by throwing an error. Hence the test case to cover invalid values.

walmik commented:
Hello @ddxtanx, I ve gone through this and I d like to thank you for addressing this issue and creating this PR! Fundamentally I like the underlying idea you have introduced but I feel a lot of improvement is required if this PR is to be merged in. I m willing to work with you to make that happen if you are interested. I ve left some comments already and will leave some more as we go along. Please do note, unit tests must be added for any new feature addition (for instance, tests should cover valid and invalid inputs to the new method).

walmik commented:
Wow that was fast! There are some lint errors though, you can see them here https://travis-ci.org/walmik/scribbletune/jobs/256685525

walmik commented on README.md:
```diff
@@ -135,6 +135,38 @@ Import the 3 MIDI files thus generated into your favorite music creation softwar
+const scribble = require("scribbletune");
+
+let clip = scribble.clip({
+	notes: ["c4", "d4", "e4", "f4"],
```
Please use single quotes here in the readme and in the code as well.

walmik commented on README.md:
````diff
@@ -135,6 +135,38 @@ Import the 3 MIDI files thus generated into your favorite music creation softwar
+scribble.midi(clip, "octave.midi");
+```
+then the clip would sound like it should!
+
````
Nice description! Thank you :)

walmik commented on src/index.js:
```diff
@@ -1 +1,19 @@
+const midi = require('./midi');
+const scales = require('./modes');
+function setMiddleC(octaveIndex){
+    if(typeof octaveIndex !== "number" || Math.floor(octaveIndex) !== octaveIndex){
```
This check should be in the main function (which is in the midi.js file).
Also here you can omit defining this function. Instead later in this file where we have `module.exports = ...`, could you check if something like this would work:
```
module.exports = {mode: scale, scale, chord: chord.getChord, listChords: chord.listChords, modes, scales: modes, clip, pattern, midi: midi.midi, setMiddleC: midi.setMiddleC};
```

walmik replied:
Once you move the validation to the function definition to midi.js, you could replace the `if` block with,
```
assert(Number.isInteger(octaveIndex), 'Octave index should be an integer');
```

walmik commented on src/midi.js:
```diff
@@ -22,19 +31,25 @@ const midi = (notes, fileName) => {
 		if (noteObj.note) {
+			var note = noteObj.note[0];
+			//Get the octave identifier (2 for a2, 5 for e5)
+			var oct = parseInt(note.slice(1,note.length));
```
Will this work in case of 2 letter notes such as Bb (B flat)

ddxtanx replied:
It does now! Notes will now transpose if they are natural, flat or sharp. I also added functionality to transpose entire chords.

ddxtanx replied:
To make sure all features are fully transposed, I also added functionality to transpose all scales generated by scribble.mode.

walmik commented:
Wow this a big change. I ll need some time to go over it.
Having said that I want to thank you for putting so much time and effort into this. Tomorrow I m gonna be speaking at [ForwardJS](https://forwardjs.com/). I ll make it a point to highlight your contribution as a token of my appreciation. 

I ll go over these changes and get back to you in a day or two.

walmik commented on src/chord.js:
```diff
@@ -110,10 +111,13 @@ const getChord = str => {
 		});
 	});
-
+	console.log(arr);
```
We can remove this console log.

walmik commented:
@ddxtanx Did you remove the tests?

ddxtanx commented:
@walmik I have not made any tests yet, but I am working on creating them now.

ddxtanx commented:
@walmik Tests have been added in.

walmik commented:
Thanks for adding the tests @ddxtanx I ll merge this in soon.
BTW I did mention your name (along with 2 other folks) at Forward JS last Thursday. I ll share the video with you once it s published.

ddxtanx commented:
@walmik Thank you for the mention! I just added the pure transpositions commit for completeness, so notes can now be truly transpose to any given octave. (Tests for this functionality were also added)

walmik commented:
@ddxtanx I m gonna merge this in now. Thanks so much for your diligence and your contributions! I ll play with this locally a bit (might tweak a bit) and if some changes are made I ll put out a PR, you can review it. Once we do all that, I ll bump the version and publish it.

ddxtanx commented:
@walmik Thank you for helping out the development process, and for merging this in!
